### ISSUE TITLE
Replace unit testing framework with mocha/chai/sinon/istanbul

### ISSUE THREAD

**Issue by JohnnyEstilles** [howardabrams/node-mocks-http#55]
## Mocha, Chai, SinonJS and Instanbul
I believe we can improve our unit testing by implementing a more flexible framework, with more readable assertion library, better isolation of the unit of code being tested, and code coverage reporting. I propose that we:

- [x] replace nodeunit with [Mocha](http://mochajs.org/)
- [x] use [Chai](http://chaijs.com/) as the assertion library
- [x] use [SinonJS](http://sinonjs.org/) for spying, mocking and stubbing
- [x] use [Istanbul](https://gotwarlost.github.io/istanbul/) for code coverage
- [x] use gulp as the task runner
- [x] add gulp tasks for running test and coverage reports
- [x] add test for http-mock
- [x] add tests for mockEventEmitter
- [x] add tests for mockWritableStream
- [x] migrate existing tests for mockRequest
- [x] migrate existing tests for mockResponse
- [x] update travis-ci to use gulp tasks
- [x] remove legacy tests
- [x] remove dependencies to legacy test

### Testing Framework: Mocha
I know this may be a matter of choice, but I believe `mocha` to be a much better javascript unit testing framework than `nodeunit`. These as just some of my reasons:

- highly extensibly
- multiple reporter options available out of the box
- multiple syntax options available out of the box
- compatible with many existing assertion libraries (I personally prefer `chai`)

### Assertion Library: Chai
Again, this is probably just a matter of style choice, but I just like the readability of Chai's assertions.

### Spy/Mock/Stub Library: SinonJS
Sinon is a great library for spying/mocking/stubbing. It is my go-to library for isolating my unit tests. I think it will be a great addition to our testing framework.

### Code Coverage: Istanbul
I like istanbul because it tracks statement, branch, and function coverage. It also includes both HTML and LCOV reporting out of the box.

---

I think we should implement these before our move to `2.0`. This will be the last minor release before starting work on `2.0`.

**Comment by howardabrams** [comment]
The original tests for this project were written in Mocha years ago, but figured I should convert the mocking project to use the built in testing framework so that other people would have a lower barrier to submit pull requests.

That said, Mocha is much better, and installing packages is pretty simple for any one who'd like to help. I have not used Chai, Sinon or Instanbul, but they look intriguing. As long as we have a simple script that can run the tests ...

Speaking of which, the original project used Jake, but I thought it was a bit complex and didn't seem like it was going any where, so I scrapped it for a simple script. With all the improvements you're making, do we want to change this at some time to use Grunt, or something similar?

Chat with you later.

**Comment by JohnnyEstilles** [comment]
@howardabrams! Wasn't aware that Mocha was the original testing framework. I guess I haven't gone back for enough into the historical commits. :-)

Running the tests will be as simple as `npm test`. Nothing out of the ordinary there. Chai and Sinon will be use within the unit tests themselves, so they will have no impact on how the tests are run. Istanbul will be run as part of the test script (probably using a task runner).

I'm glad you mentioned the implementation of a task runner. That was actually my next issue to tackle. I typically use Gulp, but if you prefer Grunt I don't mind at all. My intention is to have a number of tasks available for the contributors (e.g. lint checks, running unit tests, running unit tests with code coverage, and maybe a do-all task). I'll open a separate issue to discuss.

**Comment by howardabrams** [comment]
Great. I do not have an opinion on Grunt vs Gulp (haven't heard of that... of course, I'm so out of the Node loop right now, that doesn't surprise me).

**Comment by JohnnyEstilles** [comment]
I've already completed the test framework overhaul. The task list in my opening comment enumerates all the changes. All the work is already merged into master, and is ready to be released in `1.4`.
